### ISSUE TITLE
CNN with various Optimizer Algorithms

### ISSUE THREAD

**Issue by OppiHmiK** [OppiHmiK/Artificial-Intelligence#1]
**Develope env : Ubuntu 18.04, Python 3.5.5
Using Packages : Tensorflow, Matplotlib**

**1. CNN with Gradient Descent Optimizer**

import tensorflow as tf
import matplotlib.pyplot as plt
from tensorflow.examples.tutorials.mnist import input_data
mnist = input_data.read_data_sets('MNIST_data/', one_hot = True)

i_data = tf.placeholder("float", shape = [None, 784])
o_data = tf.placeholder("float", shape = [None, 10])
i_img = tf.reshape(i_data, [-1, 28, 28, 1])

def w_val(shape):
    init = tf.truncated_normal(shape, stddev = 0.1)
    return tf.Variable(init)

def b_val(shape):
    init = tf.constant(0.1, shape = shape)
    return tf.Variable(init)

def conv2d(i_data, weight):
    return tf.nn.conv2d(i_data, weight, strides = [1, 1, 1, 1], padding = "SAME")

def max_pool(i_data):
    return tf.nn.max_pool(i_data, ksize = [1, 2, 2, 1], strides = [1, 2, 2, 1], padding = "SAME")

w_conv1 = w_val([5, 5, 1, 32])
b_conv1 = b_val([32])

h_conv1 = tf.nn.relu(conv2d(i_img, w_conv1) + b_conv1)
h_pool1 = max_pool(h_conv1)

w_conv2 = w_val([5, 5, 32, 64])
b_conv2 = b_val([64])

h_conv2 = tf.nn.relu(conv2d(h_pool1, w_conv2) + b_conv2)
h_pool2 = max_pool(h_conv2)

w_fc1 = w_val([7*7*64, 1024])
b_fc1 = b_val([1024])

h_pool_flat = tf.reshape(h_pool2, [-1, 7*7*64])
h_fc1 = tf.nn.relu(tf.matmul(h_pool_flat, w_fc1) + b_fc1)

keep_prob = tf.placeholder("float")
h_fc1_drop = tf.nn.dropout(h_fc1, keep_prob)

w_fc2 = w_val([1024, 10])
b_fc2 = b_val([10])

o_conv = tf.nn.softmax(tf.matmul(h_fc1_drop, w_fc2) + b_fc2)
CEE = -tf.reduce_sum(o_data * tf.log(o_conv))

opti = tf.train.GradientDescentOptimizer(learning_rate=1e-4).minimize(CEE)
pred = tf.equal(tf.argmax(o_conv, 1), tf.argmax(o_data, 1))
accu = tf.reduce_mean(tf.cast(pred, "float"))
hist = {'[ Cost ]' : [], '[ Accuracy ]' : []}

with tf.Session() as sess:
    sess.run(tf.initialize_all_variables())

    for rep in range(1000):
        batch = mnist.train.next_batch(10)
        if rep % 10 == 0 or rep == 999:

            train_CEE = sess.run(CEE, feed_dict = {i_data : batch[0], o_data : batch[1], keep_prob : 1.0})
            train_accu = sess.run(accu, feed_dict = {i_data : batch[0], o_data : batch[1], keep_prob : 1.0})
            print("[ Step ] : %d \n[ Cost ] : %g, [ Accuracy ] : %g"%(rep, train_CEE, train_accu))

        sess.run(opti, feed_dict = {i_data : batch[0], o_data : batch[1], keep_prob : 0.5})
        hist['[ Cost ]'].append(sess.run(CEE, feed_dict = {i_data : batch[0], o_data : batch[1], keep_prob : 1.0}))
        hist['[ Accuracy ]'].append(sess.run(accu, feed_dict = {i_data : batch[0], o_data : batch[1], keep_prob : 1.0}))
    print("[ Cost ] : %g, [ Test Accuracy ] %g"%(sess.run(CEE, feed_dict= {i_data : mnist.test.images, o_data : mnist.test.labels, keep_prob : 1.0}),
            sess.run(accu, feed_dict = {i_data : mnist.test.images, o_data : mnist.test.labels, keep_prob : 1.0})))

graph = plt.figure()
ax_Cost = graph.add_subplot(111)
line = ax_Cost.plot(range(1000), hist['[ Cost ]'], label='Cost', color='darkred')
plt.setp(line, color = 'r')
plt.xlabel('Epochs')
plt.ylabel('Cost (color : Red)')
ax_Cost.grid(linestyle = '--', color = 'black')

ax_Accu = ax_Cost.twinx()
line2 = ax_Accu.plot(range(1000), hist['[ Accuracy ]'], label='Accuracy', color='darkred')
plt.setp(line2, color = 'b')
plt.ylabel('Accuracy (color : Blue)')
ax_Accu.yaxis.tick_right()
ax_Accu.grid(linestyle = '--', color = 'black')

plt.title("CNN with Gradient Descent Optimizer")
plt.show()
plt.legend()

![cnn with gradient descent optimizer](https://user-images.githubusercontent.com/45390469/51226493-0ec21c80-1993-11e9-9b0b-67a6d0862d4d.png)

**2. CNN with Momentum Optimizer**

...
opti = tf.train.MomentumOptimizer(learning_rate=1e-4, momentum = 0.9).minimize(CEE)
...

![cnn with momentum optimizer](https://user-images.githubusercontent.com/45390469/51226501-1681c100-1993-11e9-8c78-5802323c466f.png)

**3. CNN with AdaGrad Optimizer**

...
opti = tf.train.AdagradOptimizer(learning_rate=1e-4).minimize(CEE)
...

![cnn with adagrad optimizer](https://user-images.githubusercontent.com/45390469/51226515-1da8cf00-1993-11e9-9d98-37b725f352e5.png)

**4. CNN with Adam Optimizer**

...
opti = tf.train.AdamOptimizer(learning_rate=1e-4).minimize(CEE)
...

![cnn with adam optimizer](https://user-images.githubusercontent.com/45390469/51226528-24374680-1993-11e9-8493-46dc7498f9d5.png)
